### ISSUE TITLE
Nashorn Server Side Rendering - null is not a function in <eval> at line number 6621

### ISSUE THREAD

**Issue by ygrenzinger** [facebook/react#2427]
Hello to the React team !

I'm poking around ReactJS with server side rendering on Nashorn and I keep getting a very annoying problem.

The whole code is here: https://github.com/ygrenzinger/http-fluent-react
And the single JSX file is https://github.com/ygrenzinger/http-fluent-react/blob/master/src/main/resources/app/react/reactComponents.jsx

The faulting code is the options in the select:
```
    <div>
        <div>
            <span>Sort by</span>
            <select name="orderBy" onChange={this.handleOrderByChange}>
                /* remove option to remove this error: TypeError: null is not a function in at line number 6621 */
                <option key="orderby-price" value="price">price</option>
                <option key="orderby-variation" value="variation">variation</option>
                <option key="orderby-symbol" value="symbol">symbol</option>
                <option key="orderby-company" value="company">company</option>
            </select>
        </div>

        <div className="stocks">
        {
          rows
        }
        </div>
      </div>
```

The very sad things it works on browser but not on Java 8 - Nashorn - JS engine :'(
But **I am hoping you could have some tips so I could maybe solve it ... or see the Nashorn team if it's a bug in the JS engine**

I am using React 0.11.2 but 0.12 doesn't help.

PS: I have a branch with a more precise context with a similar problem and again option is the problem
```
var BugSelect = React.createClass({
  displayName: 'BugSelect',
  render: function() {

    var options = ['price', 'variation', 'symbol', 'company'].map(function(opt, i) {
        return <option key={i} value={opt} label={opt}>{opt}</option>;
    });

    return (
      <div>
<select value="B">
    <option value="A">Apple</option>
    <option value="B">Banana</option>
    <option value="C">Cranberry</option>
  </select>
      </div>
    );
  }
});
```

https://github.com/ygrenzinger/http-fluent-react/blob/bug-option-element/src/main/resources/app/react/reactComponents.jsx

**Comment by zertosh** [comment]
@zpao @ygrenzinger: I was/am having the same problem with `nashorn 1.8.0_25`. I put up [nashorn-react-fail](https://github.com/zertosh/nashorn-react-fail) with a minimal test case to reproduce it. I think it's a Nashorn, not a React, issue. Debugging this stuff is a pain in the ass so I don't know for sure - but I did find a "workaround".

tl;dr: Under certain conditions, when using React's context, this error happens. Adding differently named non-lifecycle methods to the Parent and the Child seems to make it work.

**Comment by zpao** [comment]
@zertosh thanks for the repro!

cc @jsfb I know you have more Java background than I do - could you look into this?

**Comment by jsfb** [comment]
@zertosh Yes, thank you, great repo.

@zpao I can confirm that it fails using npm-react 0.12.2.  I hacked in my internal build of v0.13.0-alpha, and the error went away.  That means it is likely either an issue with the npm build process, an issue which has been fixed in master, or a heisenbug bug in nashorn.  I'll do an npm-build of 0.12 and 0.13 to see what happens, and then I think we'll be able to close this one out.

**Comment by jsfb** [comment]
I shimmed in an internal build from 0.12.2, and it didn't work.  Since it sounds like it was broken for a while (at least since 0.11.2), and it appears to be resolved in 0.13, I'm going to close this bug out as resolved.

@zertosh It may be worth checking/building master, and using that version of react.  If you run into this issue again on 0.13, let's re-open the issue and we'll dig further.

**Comment by zertosh** [comment]
Thanks @jsfb, that did it! Even `0.13.0-alpha.1` is working just fine and seems stable enough, so I'm going to keep using it in one of our Nashorn apps to see if I catch anything else.
